dace/frontend/python/preprocessing.py

```diff
@@ -300,36 +309,120 @@ def has_replacement(callobj: Callable, parent_object: Optional[Any] = None, node
+            else:
+                raise NotImplementedError(f'Key type {type(kw).__name__} is not supported')
+
+        return (make_remake(kwarg_names), len(arg.keys))
+    return (None, 1)
+
+
+def flatten_callback(func: Callable, node: ast.Call):
     """
-    Creates a version of the function that has no keyword arguments and matches the number of arguments exactly.
-    Used for creating callbacks from C to Python with keyword arguments.
+    Creates a version of the function that has only marshallable arguments and no keyword arguments.
+    Arguments in callback matches the number of arguments used exactly.
+    Used for creating callbacks from C to Python with keyword arguments or other Pythonic structures 
+    (such as literal lists).
     """
-    if not node.keywords:
+
+    # Find out if any Python arguments should be flattened
+    unflatten_instructions: Dict[int, Tuple[Callable, int]] = {}
+    curarg = 0
+    for arg in node.args:
+        call, inc = _create_unflatten_instruction(arg)
+        if call is not None:
+            unflatten_instructions[curarg] = call, inc
+        curarg += inc
+    for kw in node.keywords:
+        call, inc = _create_unflatten_instruction(kw.value)
+        if call is not None:
+            unflatten_instructions[curarg] = call, inc
+        curarg += inc
+
+    # Nothing to do, early exit
+    if not node.keywords and not unflatten_instructions:
         return func
+
     keywords = [kw.arg for kw in node.keywords]
     poscount = len(node.args)
 
     # Using two levels of functions to ensure keywords are stored with the callback
-    if poscount == 0:
-        def make_cb(keywords, _):
+    if unflatten_instructions:
+        # If unflattening is necessary, have one version of the callback
+        def make_cb(keywords, poscount, instructions):
             def cb_func(*all_args):
-                kwargs = {kw: arg for kw, arg in zip(keywords, all_args)}
-                return func(**kwargs)
-            return cb_func
-    else:
-        def make_cb(keywords, poscount):
-            def cb_func(*all_args):
-                args = all_args[:poscount]
-                kwargs = {kw: arg for kw, arg in zip(keywords, all_args[poscount:])}
+                # Create an unflattened version of the original arguments
+                unflattened = []
+                skip = 0
+                for i, arg in enumerate(all_args):
+                    if skip:  # Skip processed arguments
+                        skip -= 1
+                        continue
+                    if i in instructions:
+                        unflatten, skip = instructions[i]
+                        unflattened.append(unflatten(all_args[i:i + skip]))
+                        skip -= 1
+                    else:
+                        unflattened.append(arg)
+
+                args = unflattened[:poscount]
+                kwargs = {kw: arg for kw, arg in zip(keywords, unflattened[poscount:])}
```

Comment:
I think the posargs all are before the kwargs in `all_args` so the keys of instructions are all larger than `poscount`. Couldn't this logic be simplified to: 
```suggestion
                args = all_args[:poscount]
                unflattened = []
                for i, (unflatten, skip) in sorted(instructions.items()):
                    unflattened.append(unflatten(all_args[i:i + skip]))
                kwargs = {kw: arg for kw, arg in zip(keywords, unflattened)}
```
Or maybe even shorter with more list comprehension.

Reply:
The current version with the skips and the counters looks like a weird `while`/C-style `for` emulation.

Reply:
ok, I like the loop over instructions a lot more! The unflattened/poscount is still an issue, because even positional arguments are flattened (think `print([a, b])`). I'll revise.